lib/R.sc

```diff
@@ -1244,27 +1251,26 @@ RMultiOscillatorModule : RModule {
 
 			var pulseWidth = (
 				param_PulseWidth + (sig_PWM * param_PWM)
-			// ).clip(0, 1); // TODO: remove ?
-			).linlin(0, 1, 0.05, 0.95); // TODO: ??add to other Pulse oscs too
+			).clip(0, 1);
 
 			Out.ar(
 				out_Sine,
-				SinOsc.ar(frequency) * 0.5
+				SinOsc.ar(frequency + (sig_LinFM * param_LinFM * 9000)) * 0.5
 			);
 
 			Out.ar(
 				out_Triangle,
-				LFTri.ar(frequency) * 0.5 // not band limited
+				LFTri.ar(frequency + (sig_LinFM * param_LinFM * 4000)) * 0.5 // not band limited
 			);
 
 			Out.ar(
 				out_Saw,
-				Saw.ar(frequency) * 0.5
+				Saw.ar(frequency + (sig_LinFM * param_LinFM * 500)) * 0.5
 			);
 
 			Out.ar(
 				out_Pulse,
-				Pulse.ar(frequency, pulseWidth / 2) * 0.5
+				Pulse.ar(frequency + (sig_LinFM * param_LinFM * 500), pulseWidth / 2) * 0.5
```

Comment:
I'm interested in the reason different coefficients are used per oscillator wave? Are these tuned purely by ear? (To be honest haven't yet found the time to audition all input/output variants of these modules yet)

I'm considering how these relate to the `LinFM` `ControlSpec.new(0, 5, 'lin', 0.01, 0, "")` (modulation ranges I typically set to 0-1 or -1 - 1 and then present the values as percent 0 - 100% -100% - 100%). However, in this case one way to approach this - since the FM is linear - is to have the `LinFM` directly steer the Hz delta and remove the coefficients. UI wise this could of course be presented and managed differently in the script. Thoughts?

Reply:
yea I tuned the coefficients by ear - for non-sine waves there are very different points where the waveform starts to break apart. 

I was kind of trying to figure out controlspecs while I was doing this hence some of the weirdness - In that FM test script I linked you on lines you can kind of tell what I was going for. the range is 0-2 and the idea was 1 is a comfortable, full harmony (matched by ear between waveshapes) but you have the option of pushing it further into 2. but I also didn't want to strictly limit it on the supercollider side which is why I used 5 here.

but maybe there are better was to go about this. I also couldn't figure out how to get controlspec to evenly step through a high number range which is why I'm using the 0-2 scale, which feels pretty good in it's current implementation. but maybe maybe 0-200% makes the most sense? as long as we can keep this same sort of step size feel in the controlspec I'm pretty indifferent about units

Reply:
Thanks. I haven't checked out the script yet, will do this evening and get back to you!

> I had some wierd stuff where if I changed the range to 0-3 the step size changed from 0.01 to 0.1 even though I had the step set to 0.01 - not sure what was going on and maybe I can revisit it and try again


This behavior is when you used the norns encoders? What you're experiencing is probably how the lua side of norns translates encoder delta movements for parameters based on controlspecs. It simply adjusts the delta onto the unmapped parameter value `0...1` and does not take the step defined in the controlspec into consideration. See http://doc.sccode.org/Classes/ControlSpec.html re unmapped/mapped values.

Step is applied when the raw value is mapped back and sent to the engine. It's a bit complicated - best thing would probably be to improve the norns lua parts to take step into consideration. When high resolution controllers are used (I use my arc a lot) it's not as evident even on the lua side.

Reply:
Sorry for the delay. I'm still not sure how to best define the `LinFM` spec range. My inclination right now is to have `LinFM` be specifying the exact linear Hz range of the incoming control signal (modulator). So setting 4000 would mean a `-1.0` to `1.0` signal would modulate `-4000` to `4000` Hz.

Reply:
I think that makes sense ! I think if I revisit it now I might have more success getting the controlspecs set up more sensibly and use the real values.

related question - with the way that these work is it possible/sensible to set things up so that the controlspecs in SC are higher than in Lua, so if the user wants they can increase the maxval and get a wider control? I had a problem in the synth i'm building where I wanted a longer attack time range for ADSREnv, which meant modifying SC - but obviously others may also want it shorter. here, I kinda want to set a certain range for my uses but also let others specify a larger range in Lua. not sure if there's a good way to work with that in the existing structure